master/buildbot/configurators/janitor.py

```diff
@@ -39,43 +40,67 @@ class LogChunksJanitor(BuildStep):
     name = 'LogChunksJanitor'
     renderables = ["logHorizon"]
 
-    def __init__(self, logHorizon):
+    def __init__(self, logHorizon=None, horizonPerBuilder=None):
         super().__init__()
         self.logHorizon = logHorizon
+        self.horizonPerBuilder = horizonPerBuilder
 
     @defer.inlineCallbacks
     def run(self):
-        older_than_timestamp = datetime2epoch(now() - self.logHorizon)
-        deleted = yield self.master.db.logs.deleteOldLogChunks(older_than_timestamp)
-        self.descriptionDone = ["deleted", str(deleted), "logchunks"]
+        if self.logHorizon is not None:
+            older_than_timestamp = datetime2epoch(
+                now() - self.logHorizon)
+            deleted = yield self.master.db.logs.deleteOldLogChunks(
+                older_than_timestamp=older_than_timestamp)
+            self.descriptionDone = ["deleted", str(deleted), "logchunks"]
+
+        if self.horizonPerBuilder is not None:
+            deleted = yield self.master.db.logs.deleteOldLogChunks(
+                horizonPerBuilder=self.horizonPerBuilder)
+            self.descriptionDone = ["deleted", str(deleted), "logchunks"]
+
         return SUCCESS
 
 
 class BuildDataJanitor(BuildStep):
     name = 'BuildDataJanitor'
     renderables = ["build_data_horizon"]
 
-    def __init__(self, build_data_horizon):
+    def __init__(self, build_data_horizon=None, horizonPerBuilder=None):
         super().__init__()
         self.build_data_horizon = build_data_horizon
+        self.horizonPerBuilder = horizonPerBuilder
 
     @defer.inlineCallbacks
     def run(self):
-        older_than_timestamp = datetime2epoch(now() - self.build_data_horizon)
-        deleted = yield self.master.db.build_data.deleteOldBuildData(older_than_timestamp)
-        self.descriptionDone = ["deleted", str(deleted), "build data key-value pairs"]
+        if self.build_data_horizon is not None:
+            older_than_timestamp = datetime2epoch(now() - self.build_data_horizon)
+            deleted = yield self.master.db.build_data.deleteOldBuildData(
+                older_than_timestamp=older_than_timestamp)
+            self.descriptionDone = ["deleted", str(deleted), "build data key-value pairs"]
```

Comment:
This should be just before return I think, because if both if conditions are true then we would set descriptionDone before step is actually finished.